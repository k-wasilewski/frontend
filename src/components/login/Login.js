import React, {Component} from 'react';
import '../../css/App.css';
import {connect} from "react-redux";
import {setUsername} from "../../redux/actions";
import axios from 'axios';
import {Redirect} from 'react-router-dom';

class Login extends Component {
    constructor(props) {
        super(props);

        this.state = {
            username: '',
            password: '',

            error: null
        }

        this.errorRef = React.createRef('');

        this.handleSubmit = this.handleSubmit.bind(this);
    };

    handleSubmit(e) {
        e.preventDefault();
        const that = this;

        axios.post('https://localhost:8081/login',
            {'username': this.state.username, 'password': this.state.password}
        ).then(function (response) {
            if (response.status === 200) {
                localStorage.setItem('token', response.data.token);
                that.setState({error: null});
                that.props.setUsername(that.state.username);
            } else {
                that.setState({error: 'Niepoprawna nazwa użytkownika lub hasło'});
            }
        }).catch(() => {
            that.setState({error: 'Niepoprawna nazwa użytkownika lub hasło'});
        });
    }

    usernameOnChange = (event) => {
        const username = event.target.value;
        this.setState({username: username});
    };

    passwordOnChange = (event) => {
        const password = event.target.value;
        this.setState({password: password});
    };

    render() {
        if (this.props.username!=null) return (
            <Redirect to={'/'} />
        );
        else return (
            <div className="main">
                <form onSubmit={this.handleSubmit}>
                    <div className='form'>
                        <h2>Logowanie</h2>
                        <p className='invalid' ref={this.errorRef}>{this.state.error}</p>
                        <div className='col1'>
                            <p>Użytkownik:</p>
                            <p>Hasło:</p>
                            <button>Zatwierdź</button>
                        </div>
                        <div className='col2'>
                            <p><input type='text' name='username' value= {this.state.username}
                                      id='usernameInput' onChange={this.usernameOnChange} /></p>
                            <p><input type='password' name='password' id='passwordInput'
                                      onChange={this.passwordOnChange} value= {this.state.password} /></p>
                        </div>
                    </div>
                </form>
            </div>
        );
    };
}

function mapStateToProps(state) {
    return {
        username: state.setUsernameReducer.username
    };
};

const mapDispatchToProps = {
    setUsername
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
