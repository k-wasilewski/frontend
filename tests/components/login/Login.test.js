import React from 'react';
import renderer from 'react-test-renderer';
import store from "../../../src/redux/store";
import {Provider} from "react-redux";
import Login from "../../../src/components/login/Login";
import {BrowserRouter} from "react-router-dom";
import {configure, mount} from 'enzyme';
import Adapter from "enzyme-adapter-react-16";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";

describe("Login rendering specification", () => {
    it('Login is rendered', () => {
        const component = renderer.create(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("Login functional specification", () => {
    let component;

    beforeEach(() => {
        configure({ adapter: new Adapter() });
    });

    afterEach(() => {
       component.unmount();
    });

    it('handleSubmit() sends input values username and password as a request ' +
        'to server', () => {
        const error = console.error;
        console.error = jest.fn();

        const setItem = jest.spyOn(window.localStorage.__proto__, 'setItem');
        const mockToken = 'IUOGdiygYG';
        const mockSetUsername = jest.fn();

        var mock = new MockAdapter(axios);
        const resp = {token: mockToken};
        mock.onPost('https://localhost:8081/login')
            .reply(200, resp);

        const mockData = 'mock';
        const mockEvent = {target: {value: mockData}};

        component = mount(
            <Provider store={store}>
                <Login setUsername={mockSetUsername}/>
            </Provider>
        );

        component.find('#usernameInput').simulate('change', mockEvent);
        component.find('#passwordInput').simulate('change', mockEvent);
        component.update();

        setTimeout(function () {
            expect(setItem).toHaveBeenCalledWith('token', mockToken);
            expect(mockSetUsername).toHaveBeenCalledWith(mockData);
            console.error = error;

            done();
            mock.restore();
        }, 500)
    });
});