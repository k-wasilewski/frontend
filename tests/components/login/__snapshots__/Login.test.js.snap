// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login rendering specification Login is rendered 1`] = `
<div
  className="main"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form"
    >
      <h2>
        Logowanie
      </h2>
      <p
        className="invalid"
      />
      <div
        className="col1"
      >
        <p>
          Użytkownik:
        </p>
        <p>
          Hasło:
        </p>
        <button>
          Zatwierdź
        </button>
      </div>
      <div
        className="col2"
      >
        <p>
          <input
            id="usernameInput"
            name="username"
            onChange={[Function]}
            type="text"
            value=""
          />
        </p>
        <p>
          <input
            id="passwordInput"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
        </p>
      </div>
    </div>
  </form>
</div>
`;
